packages/tpetra/core/ext/TpetraExt_TripleMatrixMultiply_def.hpp

```diff
@@ -728,6 +773,583 @@ namespace Tpetra {
+        lno_nnz_view_t Icolind;
+        const scalar_view_t Rvals = Rmat.values, Avals = Amat.values, Pvals = Pmat.values;
+        scalar_view_t Ivals;
+
+        if (!Pview.importMatrix.is_null())
+        {
+          const KCRS& Imat = Pview.importMatrix->getLocalMatrix();
+          Irowptr = Imat.graph.row_map;
+          Icolind = Imat.graph.entries;
+          Ivals = Imat.values;
+        }
+
+        // Classic csr assembly (low memory edition)
+        //
+        // mfh 27 Sep 2016: Ac_estimate_nnz does not promise an upper bound.
+        // The method loops over rows of R, and may resize after processing
+        // each row.  Chris Siefert says that this reflects experience in
+        // ML; for the non-threaded case, ML found it faster to spend less
+        // effort on estimation and risk an occasional reallocation.
+
+        size_t Acest_nnz_per_row = std::ceil(Ac_estimate_nnz(*Aview.origMatrix, *Pview.origMatrix) / m);
+
+        // mfh 27 Sep 2016: The ac_status array is an implementation detail
+        // of the local sparse matrix-matrix multiply routine.
+
+        // The status array will contain the index into colind where this entry was last deposited.
+        //   ac_status[i] <  nnz - not in the row yet
+        //   ac_status[i] >= nnz - this is the entry where you can find the data
+        // We start with this filled with INVALID's indicating that there are no entries yet.
+        // Sadly, this complicates the code due to the fact that size_t's are unsigned.
+        size_t INVALID = Teuchos::OrdinalTraits<size_t>::invalid();
+
+        // Get my node / thread info (right from openmp or parameter list)
+        size_t thread_max =  Kokkos::Compat::KokkosOpenMPWrapperNode::execution_space::concurrency();
+        if(!params.is_null()) {
+          if(params->isParameter("openmp: ltg thread max"))
+            thread_max = std::max((size_t)1,std::min(thread_max,params->get("openmp: ltg thread max",thread_max)));    
+        }
+
+        double thread_chunk = (double)(m) / thread_max;
+
+        // mfh 27 Sep 2016: Here is the local sparse matrix-matrix multiply
+        // routine.  The routine computes Ac := R * A * (P_local + P_remote).
+        //
+        // For column index Aik in row i of A, Acol2PRow[Aik] tells
+        // you whether the corresponding row of P belongs to P_local
+        // ("orig") or P_remote ("Import").
+
+        // Thread-local memory
+        Kokkos::View<u_lno_view_t*> tl_rowptr("top_rowptr", thread_max);
+        Kokkos::View<u_lno_nnz_view_t*> tl_colind("top_colind", thread_max);
+        Kokkos::View<u_scalar_view_t*> tl_values("top_values", thread_max);
+
+        // For each row of R
+        Kokkos::parallel_for("MMM::RAP::NewMatrix::ThreadLocal",range_type(0, thread_max).set_chunk_size(1),[=](const size_t tid)
+        {
+          // Thread coordination stuff
```

Comment:
If you find yourself writing really long lambdas like this, it might make sense to break them out into a functor.